File: solutions/474/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Tue Nov 16 14:03:59 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        // dp table, initialize to 0
        vector<vector<int>> dp(m+1, vector<int>(n+1, 0));
        // iterate items
        for (const auto& str : strs) {
            // calculate the weights
            int zeroNum = 0, oneNum = 0;
            for (const auto& s : str) {
                if (s == '0') zeroNum++;
                else oneNum++;
            }
            // iterate bag size
            for (int i = m; i >= zeroNum; i--) {
                for (int j = n; j >= oneNum; j--)
                    dp[i][j] = max(dp[i][j], dp[i-zeroNum][j-oneNum] + 1);
            }
        }
        return dp[m][n];
    }
};

TEST(Test474, SimpleTest) {
    vector<string> strs{
        "10","0001","111001","1","0"
    };
    const int m = 5; 
    const int n = 3;
    Solution s; 
    EXPECT_EQ(s.findMaxForm(strs, m, n), 4);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/474/main.py
from typing import List 

class Solution:
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        # initialize dp table
        dp = [[0] * (n+1) for _ in range(m+1)]
        # iterate items
        for s in strs:
            # count number of zero and one
            zeroNum, oneNum = 0, 0
            for c in s:
                if c == '0':
                    zeroNum += 1
                else:
                    oneNum += 1
            # iterate bag size
            for i in range(m, zeroNum-1, -1):
                for j in range(n, oneNum-1, -1):
                    dp[i][j] = max(dp[i][j], dp[i-zeroNum][j-oneNum]+1)
        return dp[m][n]

if __name__ == "__main__":
    strs = ["10","0001","111001","1","0"]
    m, n = 5, 3 
    s = Solution()
    assert s.findMaxForm(strs, m, n) == 4
